### ISSUE TITLE
Give visual indication that a warning was treated as error

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#29095]
When passing `/warningaserror` to Csc/Vbc, the compilers automatically promote compiler & analyzer warnings as errors. This is great, however, as we extend the number of warnings that might be promoted - it can result in confusing behavior (case in point: https://github.com/dotnet/roslyn/issues/27819#issuecomment-410566151).

C/C++ give an indication in the error text that a warning was promoted:

```
error C2220: warning treated as error - no 'object' file generated
```

I propose, that C#/Visual Basic also give a visual indication that a warning was promoted. A couple of options:

```
error CS0168: The variable 'foo' is declared but never used (warning treated as error)
error IDE0003: Name can be simplified. (warning treated as error)
```

```
error CS0168: (warning treated as error) The variable 'foo' is declared but never used
error IDE0003: (warning treated as error) Name can be simplified.
```

```
error CS0168: warning treated as error - The variable 'foo' is declared but never used
error IDE0003: warning treated as error - Name can be simplified.
```

**Comment by sharwell** [comment]
:memo: This overlaps the "developer mode" feature which is design-pending. This mode overrides Warnings as Errors to treat them instead as warnings while working in the IDE.

**Comment by CyrusNajmabadi** [comment]
FWIW, i would ike that to be configurable.  I absolutely *hate* having warnings not be errors during development.  Because i think things are fine, and i upload to jenkins to run, only to have my code fail to compile/run-tests there.  I then have to fix things, upload again, and wait for results.  It's a massive waste of time.

--

it seems like the reason people want this (when i talked to @kuhlenh) is because warn-as-error interferes with things like debugging.  But that seems to be a problem in itself that should be solved.  i.e. i should be able to ahve warn-as-errors on without that affecting my ability to debug.

**Comment by davkean** [comment]
I wouldn't gate this on the design of "developer mode". Developer mode might never be implemented, but that does not preclude this from being implemented.

**Comment by sharwell** [comment]
@davkean I'll bring this to a preliminary design review to determine if we want to treat it as a duplicate of #29123. If we do, we can always reopen it if that feature does not move forward for some reason.

**Comment by sharwell** [comment]
> it seems like the reason people want this (when i talked to @kuhlenh) is because warn-as-error interferes with things like debugging. But that seems to be a problem in itself that should be solved. i.e. i should be able to ahve warn-as-errors on without that affecting my ability to debug.

I agree that the debugging scenario can (and probably should) be fixed independently of the presentation.

I take a mostly opposite view of warnings as errors. To me, during development errors are situations where compilation cannot proceed to a result in a manner consistent with the rules of the language. "Warnings as errors" is therefore a necessarily incorrect view of what a warning is. Warnings as errors in the CI/review process makes sense from a maintainability standpoint, but it's a maintenance policy error and not a coding/language error. This is a situation where we are certain to always disagree, which is why the proposed design offers alternatives that cater to the needs of multiple user scenarios. I can't think of a presentation mechanism that meets the needs of both Cyrus and myself because we have a rather fundamental disagreement on the meaning of 'warning'.

**Comment by CyrusNajmabadi** [comment]
> I take a mostly opposite view of warnings as errors. To me, during development errors are situations where compilation cannot proceed to a result in a manner consistent with the rules of the language.

And that's definitely *not* how i see an error :)

An error just means: an error

It is an orthogonal concept to "could i produce runnable bits".  You can produce a compilation, and things can proceed in the compilation realm, while still having 'errors' that should otehrwise prevent something like CI/Review from passing.

In that regard, almost anything can be an error.  Style can be an error.  Not meeting perf gates is an error.  Not signing a CLA is an error.  But none of these errors should actually *degrade* product functionality *unnecessarily*.  

So, if hte compiler literally cannot figure out how to produce a .dll, then that's fine.  Functionality can be degraded.  But if i decided i really don't like "var", that has no bearing on my ability to use any product functionality.  I should be able to debug/profile/commit/push/analyze/test/etc.

**Comment by CyrusNajmabadi** [comment]
> I can't think of a unified presentation mechanism that meets the needs of both Cyrus and myself because we have a rather fundamental disagreement on the meaning of 'warning'.

I personally think my perspective on what a 'warning' is is objectively better.  It enables the same scenarios, while not blocking people from valuable things they want to do and without degrading the functionality of these products.

**Comment by sharwell** [comment]
During today's design review, we saw two possible interpretations of the feature request here:

* A request for distinction of warnings as errors *in the IDE*
* A request for distinction of warnings as errors *on the command line*

Based on the issue linked in the original post, we believe this is a request for the former, and as such decided to close this issue as a duplicate of #29123. 

@davkean If you are additionally interested in a command line distinction, please open a new issue requesting a command line distinction. The design meeting today was specific to the IDE behavior and the decision made in that meeting would not apply to other scenarios.

**Comment by davkean** [comment]
This is representing all cases, both command-line and IDE.

**Comment by davkean** [comment]
I've hidden the off-topic comments. This in my mind is orthogonal to developer mode.

**Comment by sharwell** [comment]
@jaredpar Re-triaging this to Area-Compilers. Area-IDE has decided not to pursue this separate from developer mode, but if a change is made to warnings presented by the compiler then they would be surfaced in the IDE indirectly.
